File: python solutions/closest_binary_search_tree_value.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def closestValue(self, root: Optional[TreeNode], target: float) -> int:
        val = root.val

        if target < val and root.left:
            left = self.closestValue(root.left, target)
            if val - target > abs(left-target): return left
            elif val - target == abs(left-target): return min(val, left)
        elif target > val and root.right:
            right = self.closestValue(root.right, target)
            if target - val > abs(right - target): return right
            elif target - val == abs(right - target): return min(val, right)

        return val

File: ruby solutions/closest_binary_search_tree_value.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {Float} target
# @return {Integer}
def closest_value(root, target)
    val = root.val

    if target < val && root.left
        left = closest_value(root.left, target)
        return left if val-target > (left-target).abs
        return [val, left].min if val-target == (left-target).abs
    elsif target > val && root.right
        right = closest_value(root.right, target)
        return right if target-val > (right-target).abs
        return [val, right].min if target-val == (right-target).abs
    end
    val
end